### PR TITLE
LDN hostname support in direct connect

### PR DESCRIPTION
**Pull request by SoRadGaming**
This PR is being opened due to an old PR with this feature takings four months to merge. Hopefully, we can get the feature added sooner.

This PR adds support for IPv6 and Hostnames.

Regex Testing:
image

Used in Test at (https://regexr.com/):
`((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)|([0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}|([a-zA-Z0-9]+(-[a-zA-Z0-9]+)*\.)+[a-zA-Z]{2,}`

Used in Yuzu Code:
`"^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$|^([0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}$|^([a-zA-Z0-9]+(-[a-zA-Z0-9]+)*\\.)+[a-zA-Z]{2,}$"`

Complied and Worked in **Yuzu**, Windows 11.

This PR fixes https://github.com/yuzu-emu/yuzu/issues/8928
This PR fixes https://github.com/yuzu-emu/yuzu/pull/8996

### CODE REVIEW

**Comment by FearlessTobi** [comment]
the regex looks good from a first glance

**Comment by liamwhite** [comment]
I'm not opposed to merging this as is. However, it's worth pointing out that this doesn't detect many possible shorthands for IPv6 addresses. Ideally, we should just let the getaddrinfo() call be used to determine whether an address is valid, rather than trying to validate it extensively here.

**Comment by SoRadGaming** [comment]
> I'm not opposed to merging this as is. However, it's worth pointing out that this doesn't detect many possible shorthands for IPv6 addresses. Ideally, we should just let the getaddrinfo() call be used to determine whether an address is valid, rather than trying to validate it extensively here.

That would be a better way of handling it, but it seems to require more than a string input to validate which is already completed when clicking connect. I have updated the IPv6 Regex in accordance with the possible addresses. One issue is that ipv4 and hostnames have a regex conflict, for example (192.168.0.1.domainname.com) is a correct hostname but not ipv4, so the ipv4 regex is useless.

File: src/yuzu/multiplayer/direct_connect.ui [hunk 1]
```diff
@@ -60,10 +60,10 @@
             <item>
              <widget class="QLineEdit" name="ip">
               <property name="toolTip">
-               <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;IPv4 address of the host&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
+               <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;IP address of the host&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
```
Comment (liamwhite) [inline src/yuzu/multiplayer/direct_connect.ui:63]: probably should be `Address of the host`

Comment (SoRadGaming) [reply src/yuzu/multiplayer/direct_connect.ui]: Changed it to `Server Address of the host`.

**Comment by SoRadGaming** [comment]
Updated the UI to the Image Below
image

**Comment by german77** [comment]
Clang format is still failing

File: src/yuzu/multiplayer/direct_connect.ui [hunk 2]
```diff
@@ -85,6 +76,18 @@
...
+                <horstretch>0</horstretch>
+                <verstretch>0</verstretch>
+               </sizepolicy>
+              </property>
+              <property name="maximumSize">
+               <size>
+                  <width>65</width>
+                  <height>50</height>
+               </size>
+              </property>
```
Comment (Morph1984) [inline src/yuzu/multiplayer/direct_connect.ui:90]: was this meant to be added

Comment (SoRadGaming) [reply src/yuzu/multiplayer/direct_connect.ui]: `<property name="sizePolicy">
               <sizepolicy hsizetype="Fixed" vsizetype="Fixed">
                <horstretch>0</horstretch>
                <verstretch>0</verstretch>
               </sizepolicy>
              </property>`

The size Policy is useless and was used in testing UI changes, I will remove it. As for **maximumSize** property, I demonstrate below why I added it, as the port can only have a limit of 5 characters there is no need to stretch that box.

With **maximumSize**:
image
With **maximumSize** Stretched:
image

Without **maximumSize**:
image
Without **maximumSize** Stretched:
image

**Comment by liamwhite** [comment]
Could you squash these commits together? I will merge ASAP

**Comment by SoRadGaming** [comment]
> Could you squash these commits together? I will merge ASAP

Sure, I Just Squashed it all into one Commit.

**Review by liamwhite — APPROVED** [review]

**Pull request merged**